src/app/Shared/Services/Services.tsx

```diff
@@ -44,19 +44,22 @@ import { NotificationChannel } from './NotificationChannel.service';
 import { ReportService } from './Report.service';
 import { SettingsService } from './Settings.service';
 import { LoginService } from './Login.service';
+import { JmxCredentials } from './JmxCredentials.service';
 
 export interface Services {
   target: TargetService;
   targets: TargetsService;
   api: ApiService;
+  jmxCredentials: JmxCredentials;
   notificationChannel: NotificationChannel;
   reports: ReportService;
   settings: SettingsService;
   login: LoginService;
 }
 
 const settings = new SettingsService();
-const login = new LoginService(TargetInstance, settings);
+const jmxCredentials = new JmxCredentials(() => api);
```

Comment:
Should it be defined after defining `api` (just to avoid surprises)? Interestingly, JS/TS allows this.

Reply:
It can't be, because the `login` has a direct reference to `jmxCredentials`, and `api` has a direct reference to `login`. I used `() => api` as a forward-reference to avoid the circular dependency problem, similar to what you would see with `Lazy` in Dagger terms on the backend.

Reply:
Wow, interesting! sorry didn't notice the below lines. Thanks for explaining again :D

Reply:
> It can't be, because the `login` has a direct reference to `jmxCredentials`, and `api` has a direct reference to `login`. I used `() => api` as a forward-reference to avoid the circular dependency problem, similar to what you would see with `Lazy` in Dagger terms on the backend.

To solve the circular dependency problem on the backend I've used Provider<>, are there merits for using either?

Reply:
Some details on the difference here: https://dagger.dev/api/2.0/dagger/Lazy.html

In short, for the use cases you've used it on the backend it probably works out the same. If you're injecting something using `Provider` then that something gets instantiated right away, before `provider.get()` is called. `Lazy` will only perform the instantiation when `lazy.get()` is called (hence the name), so if nothing ever calls `lazy.get()` then the "something" is never instantiated. But on the backend, anything you're injecting with either a Provider or a Lazy is also a `Singleton`, and `.get()` will be called, so either one should work essentially equivalently. If the `new Something()` threw an exception then I suppose you'd get different stack traces for a Provider vs Lazy.

All that said, this example of `() => api` is in fact more akin to a Dagger Provider than a Lazy :-)